plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs

```diff
@@ -56,6 +84,23 @@ data ModelOneArgument =
 instance Default ModelOneArgument where
     def = ModelOneArgumentConstantCost 0
 
+{- Note [runCostingFun* API]
+Costing functions take unlifted values, compute the 'ExMemory' of each of them and then invoke
+the corresponding @run*Model@ over the computed 'ExMemory's. The reason why we don't just make the
+costing functions take 'ExMemory's in the first place is that this would move the burden of
+computing the 'ExMemory's onto the caller, i.e. the user defining the meaning of a builtin and it
+would be just another hoop to jump through and a completely unnecessary complication for the user.
+
+The reason why costing functions take unlifted values are:
+
+1. we need to unlift them anyway to compute the result of a builtin application, so since we already
+   need them elsewhere, we can utilize them in the costing machinery too. Otherwise the costing
+   machinery would need to do some unlifting itself, which would be wasteful
+2. the costing function might actually depend on the constants that get fed to the builtin.
+   For example, checking equality of integers stored in a 'Data' object potentially has a different
+   complexity to checking equality of lists of bytestrings
+-}
+
 {- Note [Optimizations of runCostingFun*]
 We optimize all @runCostingFun*@ functions in the same way:
 
```

Comment:
Just below here it says "the two @run*Model@ functions are called ...". Since you're working on this file, could you delete the word "two"?  I think it must be left over from some time ago: we have six of those functions now.

Reply:
> Since you're working on this file, could you delete the word "two"?

Will do.

Reply:
> Just below here it says "the two @run*Model@ functions are called ...". Since you're working on this file, could you delete the word "two"? I think it must be left over from some time ago: we have six of those functions now.

Err, it was a bad wording. E.g. in 

```haskell
runCostingFunOneArgument
    :: ExMemoryUsage a1
    => CostingFun ModelOneArgument
    -> a1
    -> ExBudget
runCostingFunOneArgument (CostingFun cpu mem) =
    case (runOneArgumentModel cpu, runOneArgumentModel mem) of
        (!runCpu, !runMem) -> onMemoryUsages $ \mem1 ->
            ExBudget (ExCPU    $ runCpu mem1)
                     (ExMemory $ runMem mem1)
```

there are two calls to `runOneArgumentModel`. And it's the same for every other `runCostingFun*` function, i.e. there are actually two calls to `run*Model` in each of them.

I changed the wording to:

```
1. the two calls to @run*Model@ are placed right after matching on the first argument, so that
   they are partially computed and cached, which results in them being called only once per builtin
```

Thanks for spotting.